### PR TITLE
C# nullability

### PR DESCRIPTION
**Pull request by NN---**
Steps:
1. Add `?` and `!` where appropriate
2. Add framework attributes in addition to JetBrains Annotations. We must not remove the annotations for older frameworks and VS support.

### CODE REVIEW

File: CodeJam.Main/Expressions/ExpressionExtensions.cs [hunk 1]
```diff
@@ -1054,7 +1054,7 @@ public static T Transform<T>([CanBeNull] this T expr, [NotNull] Func<Expression,
 		/// <returns>Modified expression.</returns>
 		[Pure, CanBeNull]
 		[ContractAnnotation("expr: null => null; expr: notnull => notnull")]
-		public static Expression Transform([CanBeNull] this Expression expr, [NotNull, InstantHandle] Func<Expression, Expression> func)
+		public static Expression Transform([CanBeNull] this Expression? expr, [NotNull, InstantHandle] Func<Expression, Expression>? func)
```
Comment (ig-sinicyn) [inline CodeJam.Main/Expressions/ExpressionExtensions.cs:1057]: Shouldn't it be `public static Expression?`

File: CodeJam.Main/Expressions/ExpressionExtensions.cs [hunk 2]
```diff
@@ -1063,7 +1063,7 @@ public static Expression Transform([CanBeNull] this Expression expr, [NotNull, I
 
 		[CanBeNull]
 		[ContractAnnotation("expr: null => null; expr: notnull => notnull")]
-		private static Expression TransformInternal([CanBeNull] this Expression expr, [NotNull, InstantHandle] Func<Expression, Expression> func)
+		private static Expression TransformInternal([CanBeNull] this Expression? expr, [NotNull, InstantHandle] Func<Expression, Expression>? func)
```
Comment (ig-sinicyn) [inline CodeJam.Main/Expressions/ExpressionExtensions.cs:1066]: Shouldn't it be `public static Expression?`

**Review by ig-sinicyn — APPROVED** [review]
Hooray! Entire net 5.0 story is a huge and astonishing work, @NN--- , thanks a lot for a great job!:)

I see no major issues in the PR, but I'm not sure if we should keep JB null/notnull annotations.
As far as I can see MS nullable annotations do not depend on target FW.
ReSharper/Rider do support MS nullable annotations.
Taken all above, JB nullable annotations are reduntant and may be removed.
We may do it in its own PR to be able to revert the deletion if we change our minds sometime later.

@andrewvk , what do you think?

**Comment by NN---** [comment]
Thanks.
The PR is not ready yet. There is some work to do with annotating precisely code to reduce amount of warnings.
IMO if we are going with nullable , the best way is to set all nullable warnings as errors.

It is correct that attributes are not dependent on framework.
I added for the old frameworks in the library itself. Need to check whether it works the same way as JB.
I do think we must keep both since there could be enough people with ReSharper 2019 without good nullable support.
ReSharper got good support for C#nullable only starting 2020.2 and still it doesn't work 100% as expected .

File: CodeJam.Main/Ranges/[Boundaries]/RangeBoundaryKind.cs [hunk 3]
```diff
@@ -14,19 +14,19 @@ internal enum RangeBoundaryKindOrdering : byte
 		/// <summary>Negative infinity,        '(-∞,??',  no From limit.</summary>
 		NegativeInfinity = 0x1,
 
-		/// <summary>Less than value,          '??,b)',   x &lt; b.</summary>
```
[the remark this answers is not in the document]
Comment (NN---) [reply CodeJam.Main/Ranges/[Boundaries]/RangeBoundaryKind.cs:17]: Agree

File: CodeJam.Main/Expressions/ExpressionExtensions.cs [hunk 4]
```diff
@@ -1038,7 +1038,7 @@ private static IEnumerable<T> TransformInternal<T>([NotNull] ICollection<T> sour
 		/// <returns>Modified expression.</returns>
 		[Pure, CanBeNull]
 		[ContractAnnotation("expr: null => null; expr: notnull => notnull")]
-		public static T Transform<T>([CanBeNull] this T expr, [NotNull] Func<Expression, Expression> func)
+		public static T Transform<T>([CanBeNull] this T? expr, [NotNull] Func<Expression, Expression>? func)
```
[the remark this answers is not in the document]
Comment (NN---) [reply CodeJam.Main/Expressions/ExpressionExtensions.cs:1041]: Yes. I made replacements using regex match. Not perfect but does the 80-20 principle ;)

**Comment by ig-sinicyn** [comment]
> I do think we must keep both since there could be enough people with ReSharper 2019 without good nullable support.

Nice argument and I'm going to strongly agree with you.

Can I help in the warnings battle? We can split the work, if you wish.

File: CodeJam.Main/Arithmetic/Operators.generated.cs [hunk 5]
```diff
@@ -45,104 +46,104 @@ private static class UnaryMinusHelper
 		#region Comparison
 		/// <summary>Gets an equality comparison function, such as (a == b) in C#.</summary>
 		[NotNull]
-		public static Func<T, T, bool> AreEqual => AreEqualHelper.LazyValue.Value;
+		public static Func<T?, T?, bool> AreEqual => AreEqualHelper.LazyValue.Value;
```
[the remark this answers is not in the document]
Comment (NN---) [reply CodeJam.Main/Arithmetic/Operators.generated.cs:49]: It is correct.
Without nullability T is either U : notnull or null if T.

With null ability enabled it order to preserve old type we must use T?.

**Comment by NN---** [comment]
Sure, you can help 👍
First I am going to make tests pass and then will go back to nullable.  

You can commit all your suggestions directly to the branch.
Just make small fixup commits and regular pull so we don't get out of sync.

File: CodeJam.Main/Ranges/[Boundaries]/RangeBoundaryKind.cs (hunk 3 above)
[the remark this answers is not in the document]
Comment (ig-sinicyn) [reply CodeJam.Main/Ranges/[Boundaries]/RangeBoundaryKind.cs]: Fxd

**Comment by ig-sinicyn** [comment]
I've added some missing nullable annotations. 

I'm going to start fixing nullability warnings on evenings next week.  It looks like a pretty good amount of work so it'll take some time.

**Comment by NN---** [comment]
What do you say about errors for all nullable warnings ?
I do this in my project and this is good but sometimes annoying :)

**Comment by ig-sinicyn** [comment]
I'd fix all warnings at first. Current analysers (both c# and ReSharper ones) may produce false positives sometimes.

As example, 
```
class Helpers<T> 
{
   public void DefaultArg([MaybeNull] T arg) {}
   public void Call() => DefaultArg(default); // < warning here
}
```

**Comment by NN---** [comment]
I use convention. JBNotNull and FWNotNull.
Do you think a better convention is to use fully qualified name instead ?

**Comment by ig-sinicyn** [comment]
> I use convention. JBNotNull and FWNotNull.

I'm fine with it and I've used it in my commits. The only thing, I've renamed FWNotNull to FwNotNull, as by .net naming guidelines.
We always can do find-replace if there will be any objections:)

**Comment by NN---** [comment]
MS doesn’t follow own convention which says 2 letters is okay for acronym.
So you have System.IO and IOException and DBNull but then they introduce ClientId:)

My point is whatever looks better we shall use, the convention contradicts itself with too many use cases.

**Comment by NN---** [comment]
My opinion that JB looks better than Jb , same for FW.
Also maybe one letter can be good enough for us?

**Comment by ig-sinicyn** [comment]
>My opinion that JB looks better than Jb , same for FW.

Okay, let's use FW then. I'm not going to commit anything until tomorrow so feel free to rename it back:)

**Comment by andrewvk** [comment]
Have you seen branch /feature/nullable_v3?

**Comment by NN---** [comment]
Yes. I saw.
I will take a closer look later when I have time.
There are several changes which are not good like removing JetBrains attributes. They must be preserved.

Do you suggest to take your branch as base one ?

**Comment by andrewvk** [comment]
I think that supporting old R# attributes is a double job, and I don't think that there are many people forced to use a Resharper versions from years ago.
As for the branch, there are 620 warnings in the old one and 3700 warnings in the current, so with the first work should be done less.

**Comment by NN---** [comment]
Not everyone is updated to the latest VS2019 16.8 and there are developers even not using VS2017.
I don't see a big problem supporting both.
Happily in current job we update every release and can easily use newer .NET Core version, but in the previous one, for instance, we couldn't afford that.

**Comment by andrewvk** [comment]
NRT support does not require a VS 16.8, as NRT has been added in C# 8.0/VS 16.3, not 9.0. And NRT works well even with FW 2.0 target (with Theraot, of course, to add System.Diagnostic.Annotation attributes). The only reason to use old attributes is to support old versions of R#/Rider that do not support C# 8.

**Comment by NN---** [comment]
There were issues with ReSharper even 2020.1 with NRT.
Only the upcoming 2020.3 has solved  issues I have discovered so far.

**Pull request closed without merging**